### ISSUE TITLE
Interfaces can return empty results

### ISSUE THREAD

**Issue by leebenson** [async-graphql/async-graphql#301]
**TL;DR - If no fields being requested in an interface query match, the empty set should be omitted.** 

In [Vector](https://vector.dev), we have an interface named `Resource` that has concrete implementations as `Source`, `Transform`, and `Sink`.

When querying for a resource via an interface, we'd like for _just_ the type being spread on to be returned. So something like this:

```gql
{
  resources {
    ... on Source {
      name
      outputType
    }
  }
}
```

Would return:

```json
{
  "data": {
    "resources": [
      {
        "name": "generator",
        "outputType": "LOG"
      },
    ]
  }
}
```

Instead, currently ALL resources are being returned. Since the spread here is over the `Source` type only, this means we get results like this:

```json
{
  "data": {
    "resources": [
      {
        "name": "generator"
      },
      {},
      {},
      {},
      {},
      {}
    ]
  }
}
```

Notice the blank `{}`. These are for implementors of the `Resource` interface that should not be returned, since they don't match any fields.

It'd be ideal to filter those out since the client never requested them.

**Comment by sunli829** [comment]
> **TL;DR - If no fields being requested in an interface query match, the empty set should be omitted.**
> 
> In [Vector](https://vector.dev), we have an interface named `Resource` that has concrete implementations as `Source`, `Transform`, and `Sink`.
> 
> When querying for a resource via an interface, we'd like for _just_ the type being spread on to be returned. So something like this:
> 
> ```graphql
> {
>   resources {
>     ... on Source {
>       name
>       outputType
>     }
>   }
> }
> ```
> 
> Would return:
> 
> ```json
> {
>   "data": {
>     "resources": [
>       {
>         "name": "generator",
>         "outputType": "LOG"
>       },
>     ]
>   }
> }
> ```
> 
> Instead, currently ALL resources are being returned. Since the spread here is over the `Source` type only, this means we get results like this:
> 
> ```json
> {
>   "data": {
>     "resources": [
>       {
>         "name": "generator"
>       },
>       {},
>       {},
>       {},
>       {},
>       {}
>     ]
>   }
> }
> ```
> 
> Notice the blank `{}`. These are for implementors of the `Resource` interface that should not be returned, since they don't match any fields.
> 
> It'd be ideal to filter those out since the client never requested them.

Let me see what is the reason, but I will fix this problem in the 2.0.0-alpha version. We have done a lot of optimization and refactoring, and it will be released soon. 🙂

**Comment by leebenson** [comment]
Thanks @sunli829 !

**Comment by sunli829** [comment]
I checked the GraphQL specification and it doesn't seem to stipulate that if the fragments in the `List` cannot be applied, no results will be returned.
So the current result should be correct?

**Comment by leebenson** [comment]
👍 I checked the spec too this morning, and couldn't see a particular mention either. I'm pretty sure in other implementations, I've come across empty results being filtered out. I'll whip up a quick demo in Apollo and see how they handle it. Will report back soon.

**Comment by sunli829** [comment]
> 👍 I checked the spec too this morning, and couldn't see a particular mention either. I'm pretty sure in other implementations, I've come across empty results being filtered out. I'll whip up a quick demo in Apollo and see how they handle it. Will report back soon.

Thank you！😃

**Comment by sunli829** [comment]
@leebenson Is this correct? 😁

**Comment by leebenson** [comment]
@sunli829 - I can confirm Apollo Server works exactly the same way. 

I'm a little surprised the GraphQL spec allows empty object results, but what you're doing right now does indeed look right 👍

This is the JS test code I used:

```js
const { ApolloServer, gql } = require("apollo-server");

const typeDefs = gql`
  interface Test {
    id: String!
  }

  type Type1 implements Test {
    id: String!
    type1: String!
  }

  type Type2 implements Test {
    id: String!
    type2: String!
  }

  type Query {
    test: [Test!]!
  }
`;

const resolvers = {
  Test: {
    __resolveType(obj) {
      if (obj.type1) {
        return "Type1";
      }

      if (obj.type2) {
        return "Type2";
      }

      return null;
    },
  },
  Query: {
    test: () => [
      {
        id: "test1",
        type1: "Should only show if Type1",
      },
      {
        id: "test2",
        type2: "Should only show if Type2",
      },
    ],
  },
};

const server = new ApolloServer({ typeDefs, resolvers });

server.listen().then(({ url }) => {
  console.log(`Server: ${url}`);
});
```

And the playground result:

<img width="796" alt="Screenshot 2020-10-14 at 21 02 42" src="https://user-images.githubusercontent.com/719722/96039529-b3bba300-0e60-11eb-964e-f8fd1ed34b85.png">

Apologies for the false issue! Closing.

**Comment by sunli829** [comment]
Thanks! 😁
